### DISCUSSION TITLE
Native patchbay with customizable number of channels

### DISCUSSION THREAD

**Discussion by messmerd** [falkTX/Carla#1952] [Ideas]
Hello, I'm working on adding support in LMMS for audio plugins with multiple audio channels. I'd like to allow LMMS's built-in Carla patchbay to support however many audio channels to/from LMMS that a user wants to expose, but Carla unfortunately only seems to support a fixed number of channels.

The C API provides `carla_get_native_patchbay_plugin` which creates a 2-input/2-output patchbay, and there are other similar functions that create a patchbay with a different fixed number of inputs and outputs, but I haven't found a way to create a patchbay with an arbitrary number of channels and be able to change that number afterwards.

In the source code, those C API functions are implemented in [source/backend/engine/CarlaEngineNative.cpp](https://github.com/falkTX/Carla/tree/17000e7fe99459b25a50094a8b00bdfa12f2bfbc/source/backend/engine/CarlaEngineNative.cpp#L1544). It appears that adding a new API function that creates a patchbay with a custom number of audio, midi, or CV channels would be a relatively simple task.

However, the ability to change the number of channels afterwards would be more difficult and I do not know enough about the codebase to say how feasible it would be.

Are there any plans to add this support?

**Comment by messmerd** [comment]
@falkTX Do you have any thoughts about this?

**Reply by falkTX** [reply]
dynamic IO is not really something I have interest in, another custom but static value is okay, fully dynamic not really

**Reply by messmerd** [reply]
That's okay. I may be able to offer a fixed set of input/output configurations in LMMS (i.e. 2x2, 3x2, 16x16), and just restart Carla in order to switch between them dynamically.

**Reply by jonandtice** [reply]
> dynamic IO is not really something I have interest in, another custom but static value is okay, fully dynamic not really

Kind of related to this but in the standalone app, I prefer using carla-patchbay or carla-rack to keep the canvas nice and clean but I think it could be better if the inputs and outputs could be added and configured with custom labeling (is that dynamic IO?). @falkTX is this a feature that could be added?

I'd prefer to use carla-patchbay and switch between 'Show Internal' and 'Show External' rather than using carla-jack-multi and adding a 16x16 patchbay plugin. Also, a single midi input is a limitation.
